hackernews.el

```diff
@@ -353,19 +437,34 @@ N defaults to 1."
 
 (defun hackernews-browse-url-action (button)
   "Pass URL of BUTTON to `browse-url'."
-  (browse-url (button-get button 'shr-url)))
+  (flet ((bget (prop) (button-get button prop)))
+    (let ((id (bget 'id))
+	  (url (bget 'shr-url))
+	  (button-type (bget 'type)))
+      (button-type-put button-type 'hackernews-visited-ids (cons id (button-type-get button-type 'hackernews-visited-ids)))
+      (if hackernews-show-visited-links
+	  (let ((inhibit-read-only t))
+	    (button-put button 'type 'hackernews-link-visited)))
+      (browse-url url))))
```

Comment:
1. `flet` is defined in `cl.el`, which is not loaded by default (and whose loading at runtime is discouraged). It is an obsolete variant of the similar `cl-flet` in `cl-lib.el`; see [`(cl) Organization`](https://www.gnu.org/software/emacs/manual/html_node/cl/Organization.html) and [`(cl) Obsolete Macros`](https://www.gnu.org/software/emacs/manual/html_node/cl/Obsolete-Macros.html). Since `cl-lib.el` was only added to Emacs 24.3, I'd rather avoid depending on it for such a trivial use-case.
2. Again, indentation should follow the project's `indent-tabs-mode` setting.
3. No need to modify the buffer's contents if `button` is already of the visited variety.
4. This hard-codes `hackernews-link-visited` even for buttons of type `hackernews-comment-count`.
5. The set of IDs should be an associative data type such as a hash table so that we can associate other metadata with each ID.

Given my comments above, I think this can be written as:
```el
(defun hackernews-browse-url-action (button)
  "Pass URL of BUTTON to `browse-url'."
  (let* ((type  (button-type button))
         (vtype (button-type-get type 'hackernews-visited-button))
         (inhibit-read-only t))
    (when (and hackernews-show-visited-links
               (not (eq type vtype)))
      (button-put button 'type vtype)
      (puthash (button-get button 'id)
               t                        ; TODO: Replace with metadata
               (cdr (assq type hackernews--visited-ids)))))
  (browse-url (button-get button 'shr-url)))
```
(Guaranteed 100% untested.)